uaparser/parser.go

```diff
@@ -231,35 +231,21 @@ func NewFromBytes(data []byte) (*Parser, error) {
 
 func (parser *Parser) Parse(line string) *Client {
 	cli := new(Client)
-	var wg sync.WaitGroup
```

Comment:
so this is my first major foray into concurrent golang but if I'm reading this right, it used to parse the user agent, device, and OS in parallel, but had to pay some extra locking cost for this, and now you are making them run serially, but only lock what they need? so the point is that we prematurely parallelized and are paying extra for that?

would it make more sense to just RLock / RUnlock once in this function?

I'm also not really sure why we need the locking... I would think the parser doesn't get mutated after it's created, other than the caching.  and the cache might have it's own locking built in?

Reply:
Right, the goroutines introduced unnecessary overhead.

> I'm also not really sure why we need the locking

I think locking is needed because `parser.RegexesDefinitions.UA` and such can be mutated when `parser.Parse()` is called concurrently. You're right about caching, it has built-in locking.

> would it make more sense to just RLock / RUnlock once in this function?

In general, it's advisable to hold locks for shorter period, so one "long" lock would cover 4 branches.

Reply:
From what I understand, `checkAndSort` mutates `parser.RegexesDefinitions.UA` when sorting it, therefore it needs a write lock. The `parser.ParseUserAgent()` iterates on `UA` and needs a read lock.

Reply:
> In general, it's advisable to hold locks for shorter period, so one "long" lock would cover 4 branches.
sure, just wondering how this balances against the cost of locking and unlocking multiple times back to back.  It seems unlikely that letting different threads interleave is very beneficial with these back to back critical sections